lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -5153,6 +5180,47 @@ ParseResult FIRCircuitParser::parseModule(CircuitOp circuit, bool isPublic,
   return success();
 }
 
+ParseResult FIRCircuitParser::parseFormal(CircuitOp circuit, unsigned indent) {
+  consumeToken(FIRToken::kw_formal);
+  StringRef id, kSpelling;
+  int64_t k = -1;
+  LocWithInfo info(getToken().getLoc(), this);
+
+  // Parse the formal operation
+  if (parseId(id, "expected a formal test name") ||
+      parseToken(FIRToken::comma, "expected ','") ||
+      parseGetSpelling(kSpelling) ||
+      parseToken(FIRToken::identifier, "expected parameter 'k' after ','") ||
+      parseToken(FIRToken::equal, "expected '=' after 'k'") ||
+      parseIntLit(k, "expected integer in k specification") ||
+      parseToken(FIRToken::colon,
+                 "expected ':' after option group definition") ||
+      info.parseOptionalInfo())
+    return failure();
+
+  // Check that the parameter is valid
+  if (kSpelling != "k" || k <= 0)
+    return emitError("Invalid parameter given to formal test: ") << kSpelling,
+           failure();
+
+  // Build out the firrtl mlir op
+  auto builder = circuit.getBodyBuilder();
+  auto formal = builder.create<firrtl::FormalOp>(info.getLoc(), id, k);
+  auto &body = formal.getBody().emplaceBlock();
+
+  // Create the statement parser
+  hw::InnerSymbolNamespace ns(circuit);
```

Comment:
```suggestion
  hw::InnerSymbolNamespace ns;
```

Maybe we can just use default constructor? I think this is a bit tricky since formal doesn't have InnerSymbolTable trait. For parsing I think it's ok to create a dummy table but we still need to check that there is no inner symbol produced within formal op